Add CUDA support to a user-created torch.futures.Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56518 Add tests for CUDAFuture
* #56516 Support custom Python classes in CUDAFuture
* **#56517 Add CUDA support to a user-created torch.futures.Future**
* #56515 Allow to specify a set of device for CUDAFuture

Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

lw committed:
Add CUDA support to a user-created torch.futures.Future

Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

[ghstack-poisoned]

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -23,6 +23,25 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
+    def __init__(self, devices: Optional[List[Union[int, str, torch.device]]] = None):
+        r"""
+        Create a ``Future`` that is allowed to use (a subset of) the given
+        devices. All devices must be CUDA devices (if PyTorch supports CUDA).
```
Curious, why do we need "if PyTorch supports CUDA" here?

lw replied:
I meant this as "if you're on a CPU-only build you cannot pass CUDA devices here". In other words, if the user is passing in a CUDA device, but we cannot create a CUDAFuture (because that code wasn't built!), then we'll throw an error.

Happy to rephrase this if you have suggestions!

mrshenli replied:
Maybe "if `torch.cuda.is_available()` returns `True`"?

mrshenli commented on torch/futures/__init__.py:
```diff
@@ -23,6 +23,25 @@ class Future(torch._C.Future, Generic[T], metaclass=_PyFutureMeta):
+        for d in devices:
+            d = torch.device(d)
+            if d.type == "cpu":
+                continue
```
The doc says all devices must be CUDA, but we actually allow CPU?

mrshenli replied:
Though we ignored it.

lw replied:
Yeah this is somewhat inconsistent. My thinking was that there is no reason to pass in a CPU device, because they're "implicit" (we'll always "synchronize" with the CPU, it's automatic), hence I didn't want to mention them in the docs to avoid creating confusion. However I didn't want to be as strict in the code to allow for some leeway.

I'm happy to address this by making the code stricter and thus making it match the docs, if that's ok?

mrshenli replied:
I am OK with either way :)

lw committed:
Update on "Add CUDA support to a user-created torch.futures.Future"


Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM

lw committed:
Update on "Add CUDA support to a user-created torch.futures.Future"


Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

[ghstack-poisoned]

lw committed:
Update on "Add CUDA support to a user-created torch.futures.Future"


Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

[ghstack-poisoned]

lw committed:
Update on "Add CUDA support to a user-created torch.futures.Future"


Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

[ghstack-poisoned]

lw committed:
Update on "Add CUDA support to a user-created torch.futures.Future"


Currently a torch.futures.Future could wrap a CUDAFuture, but it could not create one from scratch. This prevented users from using CUDAFutures in some occasions, for example when using `@rpc.functions.async_execution`, or in their own code. I don't see any reason for such a limitation, hence here I add support for this.

Differential Revision: [D27887190](https://our.internmc.facebook.com/intern/diff/D27887190/)

[ghstack-poisoned]
